src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,456 @@
+                    // "client_max_window_bits" extension parameter, the corresponding
+                    // extension negotiation response to the offer MUST NOT include the
+                    // "client_max_window_bits" extension parameter.
+                    return Err(DeflateError::Negotiation(
+                        NegotiationError::UnexpectedClientMaxWindowBits,
+                    ));
+                }
+
+                // Response with unknown parameter MUST fail the WebSocket connection.
+                _ => {
+                    return Err(DeflateError::Negotiation(NegotiationError::UnknownParameter(
+                        key.to_owned(),
+                    )));
+                }
+            }
+        }
+        Ok(DeflateContext::new(Role::Client, config))
+    }
+}
+
+// A valid `client_max_window_bits` is no value or an integer in range `[8, 15]` without leading zeros.
+// A valid `server_max_window_bits` is an integer in range `[8, 15]` without leading zeros.
+#[cfg(feature = "handshake")]
+fn is_valid_max_window_bits(bits: &str) -> bool {
+    // Note that values from `headers::SecWebSocketExtensions` is unquoted.
+    matches!(bits, "8" | "9" | "10" | "11" | "12" | "13" | "14" | "15")
+}
+
+#[cfg(all(test, feature = "handshake"))]
+mod tests {
+    use super::is_valid_max_window_bits;
+
+    #[test]
+    fn valid_max_window_bits() {
+        for bits in 8..=15 {
+            assert!(is_valid_max_window_bits(&bits.to_string()));
+        }
+    }
+
+    #[test]
+    fn invalid_max_window_bits() {
+        assert!(!is_valid_max_window_bits(""));
+        assert!(!is_valid_max_window_bits("0"));
+        assert!(!is_valid_max_window_bits("08"));
+        assert!(!is_valid_max_window_bits("+8"));
+        assert!(!is_valid_max_window_bits("-8"));
+    }
+}
+
+#[derive(Debug)]
+/// Manages per message compression using DEFLATE.
+pub struct DeflateContext {
+    role: Role,
+    config: DeflateConfig,
+    compressor: Compress,
+    decompressor: Decompress,
+}
+
+impl DeflateContext {
+    #[cfg(feature = "handshake")]
+    fn new(role: Role, config: DeflateConfig) -> Self {
+        DeflateContext {
+            role,
+            config,
+            compressor: Compress::new(config.compression, false),
+            decompressor: Decompress::new(false),
+        }
+    }
+
+    fn own_context_takeover(&self) -> bool {
+        match self.role {
+            Role::Server => !self.config.server_no_context_takeover,
+            Role::Client => !self.config.client_no_context_takeover,
+        }
+    }
+
+    fn peer_context_takeover(&self) -> bool {
+        match self.role {
+            Role::Server => !self.config.client_no_context_takeover,
+            Role::Client => !self.config.server_no_context_takeover,
+        }
+    }
+
+    // Compress the data of message.
+    pub(crate) fn compress(&mut self, data: &[u8]) -> Result<Vec<u8>, DeflateError> {
```

Comment:
To be more aligned with the lib the signature should use `Bytes` instead of `&[u8]`

```rust
fn compress(&mut self, data: Bytes) -> Result<Bytes, DeflateError>
```


It would avoid turning the `Vec<u8>` into a `Bytes` when creating the Frame objet.
For the argument data, it does not change anything, it simply forward the Bytes buffer